ArduCopter/mode_autotune.cpp

```diff
@@ -1332,44 +1396,36 @@ void Copter::ModeAutoTune::updating_d_down(float &tune_d, float tune_d_min, floa
     }
 }
 
-// updating_p_down - decrease P until we don't reach the target before time out
-// P is decreased to ensure we are not overshooting the target
-void Copter::ModeAutoTune::updating_p_down(float &tune_p, float tune_p_min, float tune_p_step_ratio, float target, float measurement_max)
+// updating_rate_p_up_d_down - increase P to ensure the target is reached while checking bounce back isn't increasing
+// P is increased until we achieve our target within a reasonable time while reducing D if bounce back increases above the threshold
+void Copter::ModeAutoTune::updating_rate_p_up_d_down(float &tune_d, float tune_d_min, float tune_d_step_ratio, float &tune_p, float tune_p_min, float tune_p_max, float tune_p_step_ratio, float rate_target, float meas_rate_min, float meas_rate_max)
 {
-    if (measurement_max < target*(1+0.5f*g.autotune_aggressiveness)) {
-        if (ignore_next == false) {
-            // if maximum measurement was lower than target so increment the success counter
-            counter++;
-        } else {
-            ignore_next = false;
-        }
-    }else{
+    if (meas_rate_max > rate_target*(1+0.5f*g.autotune_aggressiveness)) {
         // ignore the next result unless it is the same as this one
         ignore_next = true;
-        // if maximum measurement was higher than target so decrement the success counter
+        // if maximum measurement was greater than target so increment the success counter
+        counter++;
+    } else if ((meas_rate_max < rate_target) && (meas_rate_max > rate_target*(1.0f-AUTOTUNE_D_UP_DOWN_MARGIN)) && (meas_rate_max-meas_rate_min > meas_rate_max*g.autotune_aggressiveness) && (tune_d > tune_d_min)) {
```

Comment:
I think the last check should always be true (if tuning should indeed stop if `tune_d <= tune_d_min`)

Reply:
yep, that is another comment problem I will fix it.